inbound.go

```diff
@@ -211,8 +211,18 @@ func (c *Connection) dispatchInbound(_ uint32, _ uint32, call *InboundCall, fram
 
 	// TODO(prashant): This is an expensive way to check for cancellation. Use a heap for timeouts.
 	go func() {
-		if <-call.mex.ctx.Done(); call.mex.ctx.Err() == context.DeadlineExceeded {
-			call.mex.inboundTimeout()
+		select {
+		case <-call.mex.ctx.Done():
+			if call.mex.ctx.Err() == context.DeadlineExceeded {
+				call.mex.inboundTimeout()
+			}
+		// If we see an error on the exchange, then we should exit immediately
+		case connErr := <-call.mex.errCh:
+			if c.log.Enabled(LogLevelDebug) {
+				c.log.Debugf("Got an error on the exchange: %v (unblock the application)", connErr)
+			}
+			call.Response().SendSystemError(connErr)
```

Comment:
Hm, that needs more debugging then -- sending a system error after a connection is marked as failed is confusing and doesn't really make sense. I'd like to make sure we understand the root problem rather than patching it to make it work.